### ISSUE TITLE
Unexpected dead_code warning when returning Result type from main

### ISSUE THREAD

**Issue by zschreur** [rust-lang/rust#123418]
I tried this code:

```rust
#[derive(Debug)]
enum MyError {
    IOErr(std::io::Error),
}

fn main() -> Result<(), MyError> {
    let io_err = std::io::Error::new(std::io::ErrorKind::Other, "Some IO Error");
    Err(MyError::IOErr(io_err))
}
```

I expect this to behave like so:
```
$ rustc ./main.rs
$ ./main
Error: IOErr(Custom { kind: Other, error: "Some IO Error" })
```

What I get is:
```
$ rustc ./main.rs
warning: field `0` is never read
 --> ./main.rs:3:11
  |
3 |     IOErr(std::io::Error),
  |     ----- ^^^^^^^^^^^^^^
  |     |
  |     field in this variant
  |
  = note: `#[warn(dead_code)]` on by default
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
  |
3 |     IOErr(()),
  |           ~~

warning: 1 warning emitted

$ ./main
Error: IOErr(Custom { kind: Other, error: "Some IO Error" })
```

### Meta

`rustc --version --verbose`:
```
rustc 1.77.1 (7cf61ebde 2024-03-27)
binary: rustc
commit-hash: 7cf61ebde7b22796c69757901dd346d0fe70bd97
commit-date: 2024-03-27
host: x86_64-apple-darwin
release: 1.77.1
LLVM version: 17.0.6
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by zschreur** [comment]
> This is not the error reported if I run your code and you have not posted the actual error you get either.

@shepmaster I apologize. I have updated the example code and copy/pasted the output I get

**Comment by krtab** [comment]
This is the same diagnostic issue as  #123068 to which I am assigned I believe.

@long-long-float I haven't started working on it yet, so you can take both if you want.

**Comment by long-long-float** [comment]
@krtab Thank you for informing me. I want to work on the issue https://github.com/rust-lang/rust/issues/123068 in one PR or divied two PRs.

**Comment by long-long-float** [comment]
As discussed in https://github.com/rust-lang/rust/issues/88900, this behavior is intentional. However there is a bug not to show an explanation about this behavior. I fixed the bug at https://github.com/rust-lang/rust/pull/124460 and the message will be shown.

```
warning: field `0` is never read
 --> ../rust/test.rs:3:11
  |
3 |     IOErr(std::io::Error),
  |     ----- ^^^^^^^^^^^^^^
  |     |
  |     field in this variant
  |
  = note: `MyError` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis
  = note: `#[warn(dead_code)]` on by default
```
